metrics/ci_integrations/test/client/buildkite/test_utils/mock/buildkite_mock_server.dart

```diff
@@ -0,0 +1,230 @@
+          ),
+          dispatcher: _downloadArtifactResponse,
+        ),
+        RequestHandler.get(
+          pathMatcher: ExactPathMatcher(
+            '$basePath/pipelines/pipeline_slug/builds/1/artifacts/not_found/download',
+          ),
+          dispatcher: _notFoundResponse,
+        ),
+        RequestHandler.get(
+          pathMatcher: ExactPathMatcher(_downloadPath),
+          dispatcher: _downloadResponse,
+        ),
+      ];
+
+  @override
+  List<AuthCredentials> get authCredentials => const [
+        AuthCredentials(token: 'token'),
+      ];
+
+  /// Responses with a list of [BuildkiteBuild]s having the build state
+  /// specified in the [request] parameters.
+  ///
+  /// Takes the per page and the page number parameters from the [request]
+  /// and returns the per page number of [BuildkiteBuild]s from the page with
+  /// the page number.
+  Future<void> _pipelineBuildsResponse(HttpRequest request) async {
+    final state = _extractBuildState(request);
+    final perPage = _extractPerPage(request);
+    final pageNumber = _extractPage(request);
+
+    List<BuildkiteBuild> builds = _generateBuilds(state);
+
+    _setNextPageUrlHeader(request, builds.length, perPage, pageNumber);
+
+    builds = MockServerUtils.paginate(builds, perPage, pageNumber);
+
+    final _response = builds.map((build) => build.toJson()).toList();
+
+    await MockServerUtils.writeResponse(request, _response);
+  }
+
+  /// Responses with a list of [BuildkiteArtifact]s.
+  ///
+  /// Takes the per page and the page number parameters from the [request]
+  /// and returns the per page number of [BuildkiteArtifact]s from the page with
+  /// the page number.
+  Future<void> _pipelineBuildArtifactsResponse(HttpRequest request) async {
+    final perPage = _extractPerPage(request);
+    final pageNumber = _extractPage(request);
+
+    List<BuildkiteArtifact> artifacts = _generateArtifacts();
+
+    _setNextPageUrlHeader(request, artifacts.length, perPage, pageNumber);
+
+    artifacts = MockServerUtils.paginate(artifacts, perPage, pageNumber);
+
+    final _response = artifacts.map((artifact) => artifact.toJson()).toList();
+
+    await MockServerUtils.writeResponse(request, _response);
+  }
+
+  /// Redirects to the artifact download URL.
+  Future<void> _downloadArtifactResponse(HttpRequest request) async {
+    final uri = Uri.parse(url);
+
+    await request.response.redirect(
+      Uri(host: uri.host, port: uri.port, path: _downloadPath),
+      status: HttpStatus.found,
+    );
+
+    await request.response.close();
+  }
+
+  /// Returns a [Uint8List] to emulate download.
+  Future<void> _downloadResponse(HttpRequest request) async {
+    await MockServerUtils.writeResponse(request, Uint8List.fromList([]));
+  }
+
+  /// Adds a [HttpStatus.notFound] status code to the [HttpRequest.response]
+  /// and closes it.
+  Future<void> _notFoundResponse(HttpRequest request) async {
```

Comment:
Would it be great to move this method to the `MockServerUtils` class? The same for the `_downloadResponse` method.